### PR TITLE
fix: make sure shuffling is calculated when querying next epoch proposers

### PR DESCRIPTION
**Pull request by matthewkeil**
**Description**

Make pulling of next proposers async to await for shuffling calculation

**Motivation**
Error was found on 1.23 release after #7120 was merged with unstable due to change made in #7130

**Notes**
@twoeths I made `getBeaconProposersNextEpoch` async to resolve this but curious if you would prefer and async helper function instead.  This is the only place where `getBeaconProposersNextEpoch` gets called so will be an easy switch if you want to keep everything on `EpochCache` sync.  Just let me know and will fix it quickly.  Currently the proposers are cached on the `epochCtx` so i could just pass that through the free function to maintain the caching or we can just recompute on each API call if there will not be many.  Just let me know which way you prefer to go but I built it like this because it seemed like the best way forward.

### CODE REVIEW

**Comment by nflaig** [comment]
> due to change made in https://github.com/ChainSafe/lodestar/pull/7130

Didn't change anything in regard to `getBeaconProposersNextEpoch`, making it async seems fine to me

**Review by nflaig — COMMENTED** [review]
unit tests are failing https://github.com/ChainSafe/lodestar/actions/runs/11304340795/job/31442660535?pr=7156#step:5:3296

**Comment by matthewkeil** [comment]
> unit tests are failing https://github.com/ChainSafe/lodestar/actions/runs/11304340795/job/31442660535?pr=7156#step:5:3296

Pushed it right before I left for dinner. Will check when I get home and fix

**Comment by matthewkeil** [comment]
> > due to change made in https://github.com/ChainSafe/lodestar/pull/7130
> 
> Didn't change anything in regard to `getBeaconProposersNextEpoch`, making it async seems fine to me

@twoeths was pretty specific to keep EpochCache all sync. But you are right, I think it's the right fix

File: packages/state-transition/src/cache/epochCache.ts [hunk 1]
```diff
@@ -918,12 +918,20 @@ export class EpochCache {
    * balances are sampled to adjust the probability of the next selection (32 per epoch on average). So to invalidate
    * the prediction the effective of one of those 32 samples should change and change the random_byte inequality.
    */
-  getBeaconProposersNextEpoch(): ValidatorIndex[] {
+  async getBeaconProposersNextEpoch(): Promise<ValidatorIndex[]> {
     if (!this.proposersNextEpoch.computed) {
+      const shuffling = await this.shufflingCache?.get(this.nextEpoch, this.nextDecisionRoot);
```
Comment (nflaig) [inline packages/state-transition/src/cache/epochCache.ts]: I think we need to handle the case where this is invoked concurrently, otherwise we end up running shuffling calculation multiple times as `this.proposersNextEpoch.computed` update will be delayed until after Promise resolves

**Comment by nflaig** [comment]
> > > due to change made in #7130
> > 
> > 
> > Didn't change anything in regard to `getBeaconProposersNextEpoch`, making it async seems fine to me
> 
> @twoeths was pretty specific to keep EpochCache all sync. But you are right, I think it's the right fix

Makes sense in general, if we can come up with an alternative

File: packages/state-transition/src/cache/epochCache.ts (hunk 1 above)
Comment (matthewkeil) [reply packages/state-transition/src/cache/epochCache.ts]: That's actually handled in the shuffling cache. Async `get` returns the same promise from multiple calls and will resolve all at the same time. So there will be multiple settings of the resolution in the EpochCache but only one calculation. As a note this will still throw if there was no request to calculate the shuffling in an epoch transition.

**Review by nflaig — COMMENTED** [review]
Tested the branch on mainnet, issue seems resolved, leaving it up to @twoeths to approve

**Comment by matthewkeil** [comment]
> Tested the branch on mainnet, issue seems resolved, leaving it up to @twoeths to approve

yep.. its deployed on feat2

File: packages/state-transition/src/cache/epochCache.ts [hunk 2]
```diff
@@ -918,12 +918,22 @@ export class EpochCache {
...
+        throw new EpochCacheError({
+          code: EpochCacheErrorCode.NEXT_SHUFFLING_NOT_AVAILABLE,
+          epoch: this.nextEpoch,
+          decisionRoot: this.getShufflingDecisionRoot(this.nextEpoch),
+        });
+      }
       const indexes = computeProposers(
         this.config.getForkSeqAtEpoch(this.epoch + 1),
         this.proposersNextEpoch.seed,
-        this.getShufflingAtEpoch(this.nextEpoch),
```
Comment (twoeths) [inline packages/state-transition/src/cache/epochCache.ts:926]: so I suppose it failed silently due to this function. Could you inspect all the consumers of this function 
and make sure it has no issue @matthewkeil ? for beacon-node we already have `shufflingCache` so no need to query here, could be in a separate PR

File: packages/state-transition/src/cache/epochCache.ts [hunk 3]
```diff
@@ -918,12 +918,22 @@ export class EpochCache {
    * balances are sampled to adjust the probability of the next selection (32 per epoch on average). So to invalidate
    * the prediction the effective of one of those 32 samples should change and change the random_byte inequality.
    */
-  getBeaconProposersNextEpoch(): ValidatorIndex[] {
+  async getBeaconProposersNextEpoch(): Promise<ValidatorIndex[]> {
```
Comment (twoeths) [inline packages/state-transition/src/cache/epochCache.ts]: using `async` here seems the best way to go, just need to drop some comments saying that this function is technically not part of the state-transition functions, it's just an util to query next epoch's proposers

File: packages/state-transition/src/cache/epochCache.ts [hunk 4]
```diff
@@ -918,12 +918,22 @@ export class EpochCache {
...
     if (!this.proposersNextEpoch.computed) {
+      if (!this.nextShuffling) {
+        this.nextShuffling = (await this.shufflingCache?.get(this.nextEpoch, this.nextDecisionRoot)) ?? null;
+      }
+      if (!this.nextShuffling) {
+        throw new EpochCacheError({
```
Comment (twoeths) [inline packages/state-transition/src/cache/epochCache.ts]: I suppose we never reach this and this is just to make the compiler happy, drop some comments for that context

Comment (matthewkeil) [reply packages/state-transition/src/cache/epochCache.ts]: yes correct. was removed in refactor

File: packages/state-transition/src/cache/epochCache.ts (hunk 2 above)
Comment (matthewkeil) [reply packages/state-transition/src/cache/epochCache.ts:926]: > so I suppose it failed silently due to this function

yes, correct.  should be fine now after the refactor because the shuffling will be ready by time `getBeaconProposersNextEpoch` is called.

> Could you inspect all the consumers of this function and make sure it has no issue @matthewkeil ?

sure.  done and all are accounted for.  `getBeaconProposersNextEpoch` was not used anywhere before the PR nico wrote so it slipped through the cracks but all other consumers should be guarded against this issue

> for beacon-node we already have `shufflingCache` so no need to query here, could be in a separate PR

good idea!  refactored

**Review by twoeths — APPROVED** [review]
lgtm, this looks simpler and thanks for keeping state-transition sync @matthewkeil

**Review by nflaig — APPROVED** [review]
LGTM

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:
